numpyro/hmc_util.py

```diff
@@ -128,13 +127,13 @@ def init_fn(z, r):
         potential_energy, z_grad = value_and_grad(potential_fn)(z)
         return IntegratorState(z, r, potential_energy, z_grad)
 
-    def update_fn(step_size, state):
+    def update_fn(step_size, inverse_mass_matrix, state):
         """
         Single step velocity verlet.
         """
         z, r, _, z_grad = state
         r = tree_multimap(lambda r, z_grad: r - 0.5 * step_size * z_grad, r, z_grad)  # r(n+1/2)
-        r_grad = grad(kinetic_fn)(r)
+        r_grad = grad(kinetic_fn)(r, inverse_mass_matrix)
```

Comment:
I have thought that we can avoid this by defining: `kinetic_fn = partial(..., inverse_mass_matrix)` but I might have been wrong. Let's do this way first and discuss about the changes (if possible) in later PRs.

Reply:
I considered that. I think the issue there is that `kinetic_fn` will not be a pure function, and will need to passed around. However that has other issues because it will need to be a static arg, but since it will keep changing as `mass_matrix` is adapted, it will keep triggering recompilation. As such, I thought it best to keep it simple and just pass the `mass_matrix` which can be easily jitted.

Reply:
I am curious that if `kinetic_fn` is changed inside the body of a while_loop, then does it cause any issue? Right now, recursive build requires jitting functions having kinetic_fn as an arg but how about iterative build? I just want to know your opinion. There is no need to test it out in this PR. :)

Reply:
I'm not completely sure - my guess would be that since traced function applications are cached by their arguments, every time we apply partial to `kinetic_fn`, we will change the hash of this partially applied function. Since the key will be missing in the cache, it will still force recompilation. Unlike some other issues, we know that this will only happen a few times, so for large enough problems it may not be as much of an overhead. I would be very surprised and curious if this wasn't the case. :)

Reply:
I go ahead and test our hypothesis in [this notebook](https://gist.github.com/fehiepsi/2fdbdd7db13183a0b3fd9fad53287cc8). Despite that I have changed `pf`, jitted h does not know about this change and still runs fast.

Reply:
@fehiepsi - It turns out that the tracer has no issues with partial function application inside a while loop or JITing the enclosing function, as it correctly traces the applied operations and doesn't need to even store/cache the function. I think it will matter in cases like below, where you should notice recompilation ticking in and increasing the runtime. 

```python
@jax.partial(jax.jit, static_argnums=(0,))
def fn(partial_ke, r):
    ...
```

Reply:
Yes, I also think so. Currently we use this pattern (@partial(jit, ...)) for recursive build, but when memory issue is addressed for iterative build, I think we can revert back to the old behaviour without any problem.

Reply:
I still think we shouldn't liberally use partial, specially if we have to do it very often for the same function, because a failure to update the function will cause silent bugs. On the other hand, if there are only a limited number of places where we have to use this, it can help readability. We can refactor this later if that turns out to be the case.

Reply:
Yup, agree. These things can be refactored later if necessary. We should avoid making any confusion which might block us. :)